File: src/main/java/forpleuvoir/hiirosakura/client/gui/GuiConfig.java
package forpleuvoir.hiirosakura.client.gui;

import fi.dy.masa.malilib.config.IConfigBase;
import fi.dy.masa.malilib.gui.GuiBase;
import fi.dy.masa.malilib.gui.GuiConfigsBase;
import fi.dy.masa.malilib.gui.GuiTextFieldInteger;
import fi.dy.masa.malilib.gui.button.ButtonBase;
import fi.dy.masa.malilib.gui.button.ButtonGeneric;
import fi.dy.masa.malilib.gui.button.IButtonActionListener;
import fi.dy.masa.malilib.util.StringUtils;
import forpleuvoir.hiirosakura.client.HiiroSakuraClient;
import forpleuvoir.hiirosakura.client.config.Configs;
import forpleuvoir.hiirosakura.client.config.HotKeys;
import forpleuvoir.hiirosakura.client.config.TogglesHotKeys;
import forpleuvoir.hiirosakura.client.gui.event.EventScreen;
import net.minecraft.client.gui.screen.Screen;

import java.util.Collections;
import java.util.List;
import java.util.Objects;

/**
 * @author forpleuvoir
 * <p>#project_name hiirosakura
 * <p>#package forpleuvoir.hiirosakura.client.gui
 * <p>#class_name GuiConfig
 * <p>#create_time 2021/6/15 20:28
 */
public class GuiConfig extends GuiConfigsBase {
    public static ConfigGuiTab tab = ConfigGuiTab.TOGGLES;

    public GuiConfig(Screen parent){
    	this();
    	this.setParent(parent);
    }

    public GuiConfig() {
        super(10, 50, HiiroSakuraClient.MOD_ID, null, "hiirosakura.gui.title.configs");
    }

    @Override
    public void initGui() {
        if (tab == ConfigGuiTab.EVENT) {
            GuiBase.openGui(new EventScreen());
            return;
        }
        super.initGui();
        this.clearOptions();
        int x = 10;
        for (ConfigGuiTab tab : ConfigGuiTab.values()) {
            x += this.createButton(x, tab);
        }
    }

    private int createButton(int x, ConfigGuiTab tab) {
        ButtonGeneric button = new ButtonGeneric(x, 26, -1, 20, tab.getDisplayName());
        button.setEnabled(GuiConfig.tab != tab);
        this.addButton(button, new ButtonListener(tab, this));
        return button.getWidth() + 2;
    }

    @Override
    protected boolean useKeybindSearch() {
        return GuiConfig.tab == ConfigGuiTab.HOTKEYS || GuiConfig.tab == ConfigGuiTab.TOGGLES_HOTKEYS;
    }

    @Override
    protected int getConfigWidth() {
        ConfigGuiTab tab = GuiConfig.tab;
        return switch (tab) {
            case TOGGLES -> 80;
            case VALUES -> 120;
            case HOTKEYS, TOGGLES_HOTKEYS -> 200;
            default -> super.getConfigWidth();
        };
    }

    @Override
    public List<ConfigOptionWrapper> getConfigs() {
        List<? extends IConfigBase> configs;
        ConfigGuiTab tab = GuiConfig.tab;
        configs = switch (tab) {
            case TOGGLES -> Configs.Toggles.OPTIONS;
            case TOGGLES_HOTKEYS -> TogglesHotKeys.getHOTKEY_LIST();
            case VALUES -> Configs.Values.OPTIONS;
            case HOTKEYS -> HotKeys.HOTKEY_LIST;
            default -> Collections.emptyList();
        };
        return ConfigOptionWrapper.createFor(configs);
    }

    private record ButtonListener(ConfigGuiTab tab, GuiConfig parent) implements IButtonActionListener {

        @Override
        public void actionPerformedWithButton(ButtonBase button, int mouseButton) {
            GuiConfig.tab = this.tab;
            if (tab == ConfigGuiTab.EVENT) {
                GuiBase.openGui(new EventScreen());
            } else {
                this.parent.reCreateListWidget();
                Objects.requireNonNull(this.parent.getListWidget()).resetScrollbarPosition();
                this.parent.initGui();
            }
        }
    }


    public enum ConfigGuiTab {
        TOGGLES("button.config_gui.toggles"),
        TOGGLES_HOTKEYS("button.config_gui.toggles.hotkeys"),
        VALUES("button.config_gui.values"),
        HOTKEYS("button.config_gui.hotkeys"),
        EVENT("button.config_gui.event");


        private final String translationKey;

        ConfigGuiTab(String translationKey) {
            this.translationKey = String.format("%s.gui.%s", HiiroSakuraClient.MOD_ID, translationKey);
        }

        public String getDisplayName() {
            return StringUtils.translate(this.translationKey);
        }
    }
}

File: src/main/kotlin/forpleuvoir/hiirosakura/client/gui/GuiConfig.kt
package forpleuvoir.hiirosakura.client.gui

import com.google.common.collect.ImmutableList
import fi.dy.masa.malilib.config.IConfigBase
import fi.dy.masa.malilib.gui.GuiConfigsBase
import fi.dy.masa.malilib.gui.button.ButtonBase
import fi.dy.masa.malilib.gui.button.ButtonGeneric
import fi.dy.masa.malilib.gui.button.IButtonActionListener
import fi.dy.masa.malilib.gui.widgets.WidgetListConfigOptions
import fi.dy.masa.malilib.util.StringUtils
import forpleuvoir.hiirosakura.client.HiiroSakuraClient
import forpleuvoir.hiirosakura.client.config.Configs
import forpleuvoir.hiirosakura.client.config.HotKeys
import forpleuvoir.hiirosakura.client.config.TogglesHotKeys
import forpleuvoir.hiirosakura.client.gui.GuiConfig.ConfigGuiTab.*
import forpleuvoir.hiirosakura.client.gui.event.EventScreen
import forpleuvoir.hiirosakura.client.gui.task.TaskScreen
import net.minecraft.client.gui.screen.Screen
import java.util.*

/**
 * @author forpleuvoir
 *
 * #project_name hiirosakura
 *
 * #package forpleuvoir.hiirosakura.client.gui
 *
 * #class_name GuiConfig
 *
 * #create_time 2021/6/15 20:28
 */
class GuiConfig() : GuiConfigsBase(10, 50, HiiroSakuraClient.MOD_ID, null, "hiirosakura.gui.title.configs") {
	constructor(parent: Screen?) : this() {
		this.parent = parent
	}

	override fun initGui() {
		if (tab == EVENT) {
			openGui(EventScreen())
			return
		} else if (tab == TASK) {
			openGui(TaskScreen())
			return
		}
		super.initGui()
		clearOptions()
		var x = 10
		for (tab in values()) {
			x += createButton(x, tab)
		}
	}

	private fun createButton(x: Int, tab: ConfigGuiTab): Int {
		val button = ButtonGeneric(x, 26, -1, 20, tab.displayName)
		button.setEnabled(Companion.tab != tab)
		addButton(button, ButtonListener(tab, this))
		return button.width + 2
	}

	override fun useKeybindSearch(): Boolean {
		return tab == HOTKEYS || tab == TOGGLES_HOTKEYS
	}

	override fun getConfigWidth(): Int {
		return when (tab) {
			TOGGLES -> 80
			VALUES -> 80
			HOTKEYS, TOGGLES_HOTKEYS -> 80
			EVENT, TASK -> super.getConfigWidth()
		}
	}

	override fun getConfigs(): List<ConfigOptionWrapper> {
		val configs: MutableList<out IConfigBase> = when (tab) {
			TOGGLES -> Configs.Toggles.OPTIONS
			TOGGLES_HOTKEYS -> TogglesHotKeys.HOTKEY_LIST
			VALUES -> Configs.Values.OPTIONS
			HOTKEYS -> HotKeys.HOTKEY_LIST
			EVENT, TASK -> ImmutableList.of()
		}
		return ConfigOptionWrapper.createFor(configs)
	}

	private inner class ButtonListener(val tab: ConfigGuiTab, var parent: GuiConfig) : IButtonActionListener {
		override fun actionPerformedWithButton(button: ButtonBase, mouseButton: Int) {
			GuiConfig.tab = this.tab
			when (tab) {
				EVENT -> {
					openGui(EventScreen())
				}
				TASK -> {
					openGui(TaskScreen())
				}
				else -> {
					parent.reCreateListWidget()
					Objects.requireNonNull<WidgetListConfigOptions>(parent.listWidget).resetScrollbarPosition()
					parent.initGui()
				}
			}
		}
	}

	enum class ConfigGuiTab(translationKey: String) {
		TOGGLES("button.config_gui.toggles"),
		TOGGLES_HOTKEYS("button.config_gui.toggles.hotkeys"),
		VALUES("button.config_gui.values"),
		HOTKEYS("button.config_gui.hotkeys"),
		EVENT("button.config_gui.event"),
		TASK("button.config_gui.task");

		private val translationKey: String
		val displayName: String
			get() = StringUtils.translate(translationKey)

		init {
			this.translationKey = String.format("%s.gui.%s", HiiroSakuraClient.MOD_ID, translationKey)
		}
	}

	companion object {
		var tab = TOGGLES
	}
}